docs/design/specs/Memory-model.md

```diff
@@ -0,0 +1,261 @@
+
+# .NET memory model
+
+## ECMA 335 vs. .NET memory models.
+ECMA 335 standard defines a very weak memory model. After two decades the desire to have a flexible model did not result in considerable benefits due to hardware being more strict. On the other hand programming against ECMA model requires extra complexity to handle scenarios that are hard to comprehend and not possible to test.
+
+In the course of multiple releases .NET runtime implementations settled around a memory model that is a practical compromise between what can be implemented efficiently on the current hardware, while staying reasonably approachable by the developers. This document rationalizes the invariants provided and expected by the CLR runtime in its current implementation with expectation of that being carried to future releases.
+
+## Alignment
+When managed by the .NET runtime, variables of built-in primitive types are *properly aligned* according to the data type size. This applies to both heap and stack allocated memory.
+
+1-byte, 2-byte, 4-byte variables are stored at 1-byte, 2-byte, 4-byte boundary, respectively.
+8-byte variables are 8-byte aligned on 64 bit platforms.
+Native-sized integer types and pointers have alignment that matches their size on the given platform.
+
+## Atomic memory accesses.
+Memory accesses to *properly aligned* data of primitive types are always atomic. The value that is observed is always a result of complete read and write operations.
+
+Values of unmanaged pointers are treated as native integer primitive types. Memory accesses to *properly aligned* values of unmanaged pointers are atomic.
+
+Managed references are always aligned to their size on the given platform and accesses are atomic.
+
+The following methods perform atomic memory accesses regardless of the platform.<br/>
+- `System.Threading.Interlocked` methods
+- `System.Threading.Volatile` methods
+
+**Example:** `Volatile.Read<double>(ref location)` on a 32 bit platform is atomic, while an ordinary read of `location` mat not be.
```

Comment:
The location has to be sufficiently aligned for given platform. For example, it has to be 8-byte aligned on Arm 32-bit.

Reply:
I think support of `Interlocked` with wider types is the reason why we over-align longs and doubles on ARM32. That makes atomic `Volatile` also possible.  

I am not sure how to word this detail properly. Also - is overalignment a universal requirement? Does Mono overalign?  (I think it would have to, for the same reasons - to support `Interlocked`)

Reply:
*properly aligned* unless the given architecture does not require it ?

Reply:
I am not sure if overalignment by itself should be a documented/required feature though. It is just one way to make `Interlocked` work with wider than pointer type, but there could be other ways (like taking a global lock, however crude it could be, it would work).

The subtle point here is that `Volatile` and `Interlocked` will be atomic when location is managed by the runtime. (

Reply:
> Volatile and Interlocked will be atomic when location is managed by the runtime.

I like this description

Reply:
Changed to have "when the location of the variable is managed by the runtime."